### PR TITLE
Fix UnicodeDecodeError in Py2.7

### LINKED ISSUE #433: Pip compile breaks with scandir==1.4
pip-compiles fails for `scandir==1.4` dependency

```
scandir==1.4 not in cache, need to check index
Traceback (most recent call last):
  File "/www/env/bin/pip-compile", line 11, in <module>
    sys.exit(cli())
  File "/www/env/lib/python2.7/site-packages/click/core.py", line 716, in __call__
    return self.main(*args, **kwargs)
  File "/www/env/lib/python2.7/site-packages/click/core.py", line 696, in main
    rv = self.invoke(ctx)
  File "/www/env/lib/python2.7/site-packages/click/core.py", line 889, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/www/env/lib/python2.7/site-packages/click/core.py", line 534, in invoke
    return callback(*args, **kwargs)
  File "/www/env/lib/python2.7/site-packages/piptools/scripts/compile.py", line 168, in cli
    results = resolver.resolve()
  File "/www/env/lib/python2.7/site-packages/piptools/resolver.py", line 105, in resolve
    has_changed, best_matches = self._resolve_one_round()
  File "/www/env/lib/python2.7/site-packages/piptools/resolver.py", line 187, in _resolve_one_round
    for best_match in best_matches
  File "/www/env/lib/python2.7/site-packages/piptools/resolver.py", line 188, in <genexpr>
    for dep in self._iter_dependencies(best_match))
  File "/www/env/lib/python2.7/site-packages/piptools/resolver.py", line 261, in _iter_dependencies
    dependencies = self.repository.get_dependencies(ireq)
  File "/www/env/lib/python2.7/site-packages/piptools/repositories/pypi.py", line 161, in get_dependencies
    dependencies = reqset._prepare_file(self.finder, ireq)
  File "/www/env/lib/python2.7/site-packages/pip/req/req_set.py", line 587, in _prepare_file
    session=self.session, hashes=hashes)
  File "/www/env/lib/python2.7/site-packages/pip/download.py", line 810, in unpack_url
    hashes=hashes
  File "/www/env/lib/python2.7/site-packages/pip/download.py", line 653, in unpack_http_url
    unpack_file(from_path, location, content_type, link)
  File "/www/env/lib/python2.7/site-packages/pip/utils/__init__.py", line 605, in unpack_file
    untar_file(filename, location)
  File "/www/env/lib/python2.7/site-packages/pip/utils/__init__.py", line 551, in untar_file
    path = os.path.join(location, fn)
  File "/www/env/lib/python2.7/posixpath.py", line 80, in join
    path += '/' + b
UnicodeDecodeError: 'ascii' codec can't decode byte 0xc6 in position 27: ordinal not in range(128)
make: *** [install] Error 1

##### Steps to replicate

1. ...have scandir==1.4 in requirements file
2. ...
3. ...

##### Expected result
Should work
...

##### Actual result
Errors out
...

### PR DESCRIPTION
**Pull request by mhluongo**
Closes #433 and #434. Pretty sure this also addresses #224 and #439, though those need confirmation.

Thanks @byjott for the fix, hopefully we can get this merged quickly.

### CODE REVIEW

File: piptools/cache.py [hunk 1]
```diff
@@ -159,6 +160,10 @@ def _reverse_dependencies(self, cache_keys):
         """
         # First, collect all the dependencies into a sequence of (parent, child) tuples, like [('flake8', 'pep8'),
         # ('flake8', 'mccabe'), ...]
-        return lookup_table((key_from_req(Requirement.parse(dep_name)), name)
+        return lookup_table((key_from_req(parse(dep_name)), name)
```
Comment (davidovich) [inline piptools/cache.py]: I believe this is not related to the issue you are trying to fix. Please open another PR for the package extras related modification, and remove this commit.

**Comment by vphilippon** [comment]
Dealth with by #568
